website/www/site/content/en/documentation/patterns/shared-class.md

```diff
@@ -0,0 +1,187 @@
+        self._customers = gen_customers(version=self._version)
+        return WeakRefDict(self._customers)
+
+    def process(self, element):
+        attr = self._customer_lookup.get(element["customer_id"], {})
+        yield {**element, **attr}
+
+
+with beam.Pipeline() as p:
+    shared_handle = shared.Shared()
+    (
+        p
+        | beam.Create(gen_orders(ts=datetime.now().timestamp()))
+        | beam.ParDo(EnrichOrderFn(shared_handle))
+        | beam.Map(print)
+    )
+
+    logging.getLogger().setLevel(logging.INFO)
+    logging.info("Building pipeline ...")
+{{< /highlight >}}
+
+## Create a cache and update it regularly on a streaming pipeline
+
+As the batch pipeline example, the customer cache is initialized as a dictionary in the `setup` method. To update the cache, it is refreshed when its timestamp is earlier than the defined threshold value of 5 seconds. The refreshment is implemented in the `start_bundle` method. Once the cached is updated, the order records are enriched with updated customer attributes. Note that it is necessary to change the `tag` argument of the `acquire` method to reload the shared object when it gets refreshed. The current timestamp value (`current_ts`) is used in this example.
+
+{{< highlight py >}}
+import random
+import logging
+import time
+from datetime import datetime
+from uuid import uuid4
+
+import apache_beam as beam
+from apache_beam.utils import shared
+from apache_beam.transforms.periodicsequence import PeriodicImpulse
+
+
+def gen_customers(version: int, num_cust: int = 1000):
+    d = dict()
+    for r in range(num_cust):
+        d[r] = {"version": version}
+    d["timestamp"] = datetime.now().timestamp()
+    return d
+
+
+def gen_orders(e: float, num_ord: int = 5, num_cust: int = 1000):
+    orders = [
+        {
+            "order_id": str(uuid4()),
+            "customer_id": random.randrange(1, num_cust),
+            "timestamp": e,
+        }
+        for _ in range(num_ord)
+    ]
+    for o in orders:
+        yield o
+
+
+# wrapper class needed for a dictionary since it does not support weak references.
+class WeakRefDict(dict):
+    pass
+
+
+class EnrichOrderFn(beam.DoFn):
+    def __init__(self, shared_handle):
+        self._max_stale_sec = 5
+        self._version = 1
+        self._customers = {}
+        self._shared_handle = shared_handle
+
+    def setup(self):
+        # setup is a good place to initialize transient in-memory resources.
+        self._customer_lookup = self._shared_handle.acquire(
+            self.load_customers, datetime.now().timestamp()
+        )
+
+    def load_customers(self):
+        time.sleep(2)
+        self._customers = gen_customers(version=self._version)
+        return WeakRefDict(self._customers)
+
+    def start_bundle(self):
+        # update the lookup table when the time difference exceeds the threshold.
+        current_ts = datetime.now().timestamp()
+        ts_diff = current_ts - self._customers["timestamp"]
```

Comment:
Rather than using `ts_diff` which can be different for each thread (causing the cache to be reloaded several times), it would be better to use a more deterministic mapping.

For example, you could do something like:

`tag = current_ts - (current_ts % stale._max_stale_sec)`. This would ensure that you're always still within `max_stale_sec`, but you're not reloading the cache more than needed. There may be an even smarter strategy here using a second shared object (which could tell you, for example, the tag of the latest results), but I think this is probably good enough

Reply:
Can you clarify more? `ts_diff` is necessary to check whether the shared object should be refreshed or not. The tag value can be updated to be more deterministic. Would you consider the following snippet is okay?

```python
class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._max_stale_sec = 60
        self._cache_version = 0
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers, self._cache_version)
        self._cache_version +=1
 
    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers()
        return WeakRefDict(self._customers)

    def start_bundle(self):
        # Update the lookup table when the time difference exceeds the threshold.
        current_ts = datetime.now().timestamp()
        ts_diff = current_ts - self._customers["timestamp"]
        if ts_diff > self._max_stale_sec:
            self._customer_lookup = self._shared_handle.acquire(self.load_customers, self._version)
            elf._cache_version +=1

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

Reply:
This is probably a bit better, but it still has some potential issues. Imagine the following sequence of events:

```
Worker spins up DoFn1 and DoFn2
DoFn1.setup() acquires shared_handle with _cache_version = 0
DoFn2.setup() acquires shared_handle with _cache_version = 0 (so far, so good)
DoFn1.start_bundle() acquires shared_handle with _cache_version = 1 (still ok - 2 copies of cache exist, but both are actually being used)
DoFn1.start_bundle() acquires shared_handle with _cache_version = 2 (at this point, _cache_version=1 is garbage collected)
DoFn2.start_bundle acquires shared_handle with _cache_version = 1 (this doesn't make sense, it should just jump to _cache_version=2, but the versions are out of sync)
```

Since there's no guarantee that bundle requests are going to get round-robined, over time there could be drift between the various `DoFn`'s `_cache_version`s, leading to extra remote calls. Over time, it is reasonably likely that the cache will degrade to the point where we're not getting any benefit from it.

That's why I suggested still grounding the current `_cache_version` in the current time, but bucketizing it so that it still resolves to a single discrete value per bucket

Reply:
Thanks for the clarification. Would you consider the following change is acceptable?

```python
# The wrapper class is needed for a dictionary, because it does not support weak references.
class WeakRefDict(dict):
    pass

class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._max_stale_sec = 60
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(
            self.load_customers, self.set_tag()
        )
 
    def set_tag(self):
        current_ts = datetime.now().timestamp()
        return current_ts - (current_ts % self._max_stale_sec)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers(tag=self.set_tag())
        return WeakRefDict(self._customers)

    def start_bundle(self):
        # Update the lookup table when the time difference exceeds the threshold.
        if self.set_tag() > self._customers["tag"]:
            self._customer_lookup = self._shared_handle.acquire(
                self.load_customers, self.set_tag()
            )

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

Reply:
Resolved by a separate commit.